docs/plugins/pluggable_authentication.md

```diff
@@ -83,3 +83,165 @@ Since PAM requires the user's password in plaintext, iRODS relies on TLS encrypt
+
+**Step 2: `auth_agent_auth_response`**
+
+This operation requires 3 keys: `digest`, `user_name`, and `zone_name`.
+
+`user_name` and `zone_name` are the username and zone name for the user being authenticated. A user of the specified name belonging the specified zone must exist in the local zone in order for authentication to succeed.
+
+`digest` is a base64-encoded MD5 hash of the user's password. The client should construct the MD5 hash by concatenating two strings:
+
+1. The randomly-generated string from the server stored in `request_result`
+2. The user's plaintext password
+
+How the plugin gets the user's password is entirely up to the library developer. For reference, the C++ client-side authentication plugin has a few ways of obtaining the password:
+
+- The `a_pw` key in the request payload
+- Prompting the user and receiving the passwords from `stdin`
+- Reading from an irodsA file
+
+For example, if the user's password is "apass", the string to hash looks like this:
+```
+64_random_bytes_encoded_as_a_string____________________________apass
+```
+The concatenated string should be used as input to a MD5 hashing function, and the resulting byte-string should be base64-encoded. The resulting string should be stored in the `digest` key.
+
+The request payload should minimally contain the following:
+```json
+{
+    "scheme": "native",
+    "user_name": "<user_name>",
+    "zone_name": "<zone_name>",
+    "digest": "<base64-encoded string of MD5 hash>"
+}
+```
+
+The response payload from the server will be identical to the request as the server-side operation does not add or remove anything.
+
+After `auth_agent_auth_response` completes, the `RsComm` in the connected agent has all of the appropriate values set and should be ready to perform regular requests.
+
+#### PAM (`pam_password`) authentication
+
+The `pam_password` authentication plugin has one server-side operation which must be called. The inputs and outputs are described below.
+
+After this server-side operation returns, the client is expected to authenticate again using a randomly-generated iRODS native authentication password.
+
+**Step 1: `auth_agent_auth_request`**
+
+This operation requires 3 keys: `a_pw`, `user_name`, and `zone_name`.
+
+`user_name` and `zone_name` are the username and zone name for the user being authenticated. A user of the specified name belonging the specified zone must exist in the local zone in order for authentication to succeed.
+
+`a_pw` is the user's **plaintext PAM password** which is going to be checked by the configured PAM service. This is why communication between client and server requires TLS to be enabled.
+
+The request payload should minimally contain the following:
+```json
+{
+    "scheme": "pam_password",
+    "user_name": "<user_name>",
+    "zone_name": "<zone_name>",
+    "a_pw": "<plaintext PAM password>"
```

Comment:
Consider replacing the hyphens and spaces with underscores in the value for `a_pw`. Or remove the underscores from the user name and zone name values.

Reply:
Removed underscores